chore(homeserver): unify source of default config

SHAcollision opened the pull request:
This PR is an attempt to:
1. Define defaults only once
2. Fallback always to default values consistently when **some** (or **all** values) are some missing. Note that `dht_relay_nodes`, `dht_request_timeout_ms` or `dht_bootstrap_nodes` had different defaults code vs toml. They used to be `None` by default in code and `Some` in the .toml. Now they are `Some` always.

Some room for clean up but hopefully this should be more consistent and maintainable than defining defaults per-case in code.

SHAcollision commented on pubky-homeserver/src/data_directory/config_toml.rs:
```diff
@@ -1,330 +1,268 @@
         config.admin.listen_socket = SocketAddr::from(([127, 0, 0, 1], 0));
         config.pkdns.icann_domain =
             Some(Domain::from_str("localhost").expect("localhost is a valid domain"));
+        config.pkdns.dht_relay_nodes = None;
+        config.pkdns.dht_request_timeout_ms = None;
+        config.pkdns.dht_bootstrap_nodes = None;
```
These were all `None` by default before. However we DO define them on the `default.config.toml` . Do we want to define them by default for all `mainnet` homeserver? They must be `None` for test to work.
```
# List of bootstrap nodes for the DHT.
# domain:port format.
dht_bootstrap_nodes = [
    "router.bittorrent.com:6881",
    "dht.transmissionbt.com:6881",
    "dht.libtorrent.org:25401",
    "relay.pkarr.org:6881",
]
# Relay node urls for the DHT.
# Improves the availability of pkarr packets.
# If not set and no bootstrap nodes are set, the default pkarr relay nodes will be used.
dht_relay_nodes = ["https://relay.pkarr.org", "https://pkarr.pubky.org"]
# Default UDP request timeout for the DHT
dht_request_timeout_ms = 2000
```

SHAcollision commented on pubky-homeserver/src/data_directory/config_toml.rs:
```diff
@@ -1,330 +1,268 @@
-        assert_eq!(c.pkdns.dht_relay_nodes, None);
-
-        assert_eq!(c.pkdns.dht_request_timeout_ms, None);
+        assert_eq!(c.admin.admin_password, "admin");
```
We were testing `dht_bootstrap_nodes` `dht_relay_nodes` `dht_request_timeout_ms` to be `None` yet defined them as Some in the `default.config.toml` . Requires double check to make sure behaviour has not changed @SeverinAlexB

SHAcollision commented on pubky-homeserver/src/data_directory/config_toml.rs:
```diff
@@ -1,225 +1,182 @@
+});
+
+/// Helper: merge two arbitrary `toml::Value` trees (recursive)
+fn merge_toml(base: &mut toml::Value, overlay: toml::Value) {
```
Is there a better place for this?

SHAcollision commented on pubky-homeserver/config.default.toml:
```diff
@@ -40,11 +40,11 @@ public_ip = "127.0.0.1"
 # The pubky tls port in case it differs from the pubky_listen_socket port.
 # Defaults to the pubky_listen_socket port.
-public_pubky_tls_port = 6287
+# public_pubky_tls_port = 6287
```
`public_pubky_tls_port` and `public_icann_http_port`
 were being asserted to be `None` however there were `Some` in the `default.toml`

SHAcollision replied:
`None` is the default when there is no value in `.toml` so we can simply comment it out.

SHAcollision commented on pubky-homeserver/Cargo.toml:
```diff
@@ -36,6 +36,8 @@ pubky-common = { version = "0.4.0-rc.0", path = "../pubky-common" }
 serde = { version = "1.0.217", features = ["derive"] }
 tokio = { version = "1.43.0", features = ["full"] }
 toml = "0.8.20"
+serde-toml-merge = "0.3.9"
```
Opted for introducing `serde-toml-merge` (light weight, looks good) and delete the helper. But if you’d rather keep it in-tree, we can re-introduce:
```rust
fn merge_toml(base: &mut toml::Value, overlay: toml::Value) {
    use toml::Value::*;

    match (base, overlay) {
        (Table(b), Table(o)) => {
            for (k, v) in o {
                b.entry(k)
                 .and_modify(|bv| merge_toml(bv, v.clone()))
                 .or_insert(v);
            }
        }
        (Array(b), Array(o)) => *b = o,
        (b, o) => *b = o,
    }
}
```

SeverinAlexB reviewed (changes requested):
I do like because we have less code but I think the PR has a problem:
The defaults are occasional different from the sample config.

The `config.default.toml` so far was more of a sample config. The defaults were defined in the code.

Let me make an example:
The bootstrap nodes are defined like this to illustrate to the user on how to use this property. The same goes for `public_pubky_tls_port`. The illustrated properties are not defaults.

```toml
public_pubky_tls_port = 6287
dht_bootstrap_nodes = [
    "router.bittorrent.com:6881",
    "dht.transmissionbt.com:6881",
    "dht.libtorrent.org:25401",
    "relay.pkarr.org:6881",
]
```

So we might need two files: `config.sample.toml` and `config.default.toml`.

SHAcollision commented:
> The `config.default.toml` so far was more of a sample config. The defaults were defined in the code.
> 
> Let me make an example: The bootstrap nodes are defined like this to illustrate to the user on how to use this property. The same goes for `public_pubky_tls_port`. The illustrated properties were not defaults so far.
> 
> ```toml
> public_pubky_tls_port = 6287
> dht_bootstrap_nodes = [
>     "router.bittorrent.com:6881",
>     "dht.transmissionbt.com:6881",
>     "dht.libtorrent.org:25401",
>     "relay.pkarr.org:6881",
> ]
> ```
> 
> So we might need two files: `config.sample.toml` and `config.default.toml`.

Sounds good, all we need to do then is to comment out those values on the `config.default.toml` . Then they will be `None` but still serve as a reference to the user in case he wants to change them.

SHAcollision commented on pubky-homeserver/src/data_directory/config_toml.rs:
```diff
@@ -1,225 +1,156 @@
     }
 
-    /// Returns the default config with all variables commented out.
+    /// Parse a raw TOML string, overlaying it on top of the embedded defaults.
+    pub fn from_raw_str(raw: &str) -> Result<Self, ConfigReadError> {
+        // 1. Parse the embedded defaults
+        let default_val: toml::Value = DEFAULT_CONFIG
+            .parse()
+            .expect("embedded defaults invalid TOML");
+
+        // 2. Parse the user's overrides
+        let user_val: toml::Value = raw.parse()?;
+
+        // 3. Deep‐merge
+        let merged_val = serde_toml_merge::merge(default_val, user_val)
+            .map_err(|e| ConfigReadError::ConfigMergeError(e.to_string()))?;
+
+        // 4. Deserialize into our strongly typed struct (can fail with toml::de::Error)
+        Ok(merged_val.try_into()?)
+    }
```
`FromStr` is already implemented and used to convert a str .toml into a config without backfilling with the defaults. Maybe this one should be called `from_str_with_defaults()` ?

SHAcollision commented:
Added a `config.sample.toml` and moved `config.default.toml` to `/data_directory` . For what is worth I think we should only have one well documented `.toml` on the top dir that serves both as default and example. We only introduced more indirection and need for implicit knowledge to new devs by having 2 `.toml`

SHAcollision commented:
Added a separate `config.sample.toml` and moved the default configuration (`config.default.toml`) into `/data_directory`.

For what it's worth, I still believe we would benefit from having a single, well-documented `.toml` file at the top-level directory serving both as defaults and an illustrative example.

Splitting it into two files introduces extra indirection, increases cognitive load, and does require more implicit knowledge from new contributors.

SeverinAlexB reviewed (commented):
> well-documented .toml file at the top-level directory serving both as defaults and an illustrative example. 

As mentioned, defaults and sample values can be different values. A merged file would lead to weird commented out values and prevent a clean sample documentation on Github.

> We are also now mantaining two sets of comments for because one of them will be copied into ~/pubky/config.toml (fully commented out) that might be used by users for initial configuration and another one that also might be used by users for configuration (config.sample.toml) directly from github.

Why should we maintain the comments in `.pubky/config.toml`?

> Splitting it into two files introduces extra indirection, increases cognitive load, and does require more implicit knowledge from new contributors.

You have explicit defaults and samples. Not sure what's the increased cognitive load or implicit knowledge.

> However sample.toml as stand alone does need to pass tests in its entirety.

Why not just run the sample through a validating test too?



I don't see how we get a clean Github sample toml with a single file without weird commented out fields which implicitly indicate that the sample is not the default.

SHAcollision commented on pubky-homeserver/src/data_directory/config.default.toml:
```diff
@@ -40,11 +40,11 @@ public_ip = "127.0.0.1"
 
 # The pubky tls port in case it differs from the pubky_listen_socket port.
 # Defaults to the pubky_listen_socket port.
```
I'm happy to remove all comments here. But then the file we copy to `./pubky/config.toml` is meaningless and useless for an initial configuration. How do we deal with it? Do we still want it?

SeverinAlexB replied:
You copy the sample config to the data_dir, not the defaults, no?

SHAcollision replied:
I can copy the `sample.config.toml`, then we need to comment it all out. Is that correct? Do we want an uncommented version in repo and a fully commented out version when copied to the data dir?

SeverinAlexB replied:
> I can copy the sample.config.toml instead, then we need to comment it all out. Is that correct? 

Yes

> Do we want an uncommented version in repo and a fully commented out version when copied to the data dir?

Either way but I am biased towards the uncommented properties version. This way, we can read and validate the config with a test. Github also highlights the values nicely.

*Assuming you mean uncommented properties, not removing the actual comments.

SeverinAlexB replied:
> I can copy the sample.config.toml instead, then we need to comment it all out. Is that correct? Do we want an uncommented version in repo and a fully commented out version when copied to the data dir?

I suggest we keep the properties set (uncommented) in the sample.config.toml. When we write it to the data directory, we programatically comment them out. This way, 

- We can validate and test that the properties are valid.
- Github highlights the values nicely.


We could also go the sample.config.toml fully commented in Github path but then we loose the validation ability. So I don't favour that.

SHAcollision commented:
Good! With the latest commit:
- The sample config is not commented out in github for better highlights
- The sample config is commented out and placed on the directory on first start up.
- The sample config is validated on it's own test case.
- The default config has no docs anymore.

SHAcollision commented on pubky-homeserver/src/data_directory/config_toml.rs:
```diff
@@ -244,87 +170,78 @@ impl ConfigToml {
-        let c: ConfigToml = ConfigToml::default();
-
+        let c = ConfigToml::default();
         assert_eq!(c.general.signup_mode, SignupMode::TokenRequired);
         assert_eq!(c.general.user_storage_quota_mb, 0);
         assert_eq!(c.general.lmdb_backup_interval_s, 0);
         assert_eq!(
             c.drive.icann_listen_socket,
```
We test against the values we expect it is parsed into. I think it's a good a sanity check or "look twice at what you just did", of course it doesn't buy us much anyway :D  We need this test as much as we needed it before when we were testing the default `ConfigToml` object against some values and the functions that defined the `ConfigToml` defaults.
